src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +169,271 @@ func (b *dbBuffer) Write(
 		}
 
-		if !bucket.canRead() {
-			return
+		stream, err := block.Stream(ctx)
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
 
-		if !blockStart.Equal(bucket.start) {
-			return
+		segment, err := stream.Segment()
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
 
-		// We need to merge all the bootstrapped blocks / encoders into a single stream for
-		// the sake of being able to persist it to disk as a single encoded stream.
-		_, err = bucket.merge()
+		checksum, err := block.Checksum()
 		if err != nil {
-			return
+			return FlushOutcomeErr, err
 		}
 
-		// This operation is safe because all of the underlying resources will respect the
-		// lifecycle of the context in one way or another. The "bootstrapped blocks" that
-		// we stream from will mark their internal context as dependent on that of the passed
-		// context, and the Encoder's that we stream from actually perform a data copy and
-		// don't share a reference.
-		streams := bucket.streams(ctx)
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
+		err = persistFn(id, tags, segment, checksum)
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	if bucket, exists := buckets.WritableBucket(WarmWrite); exists {
+		bucket.SetVersion(version)
+	}
 
-	return res, err
+	return FlushOutcomeFlushedToDisk, nil
 }
 
 func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xio.BlockReader {
 	// TODO(r): pool these results arrays
 	var res [][]xio.BlockReader
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
-		}
-		if !start.Before(bucket.start.Add(b.blockSize)) {
-			return
-		}
-		if !bucket.start.Before(end) {
+
+	b.forEachBucketAsc(func(bv BufferBucketVersions) {
+		if !bv.StartTime().Before(end) || !start.Before(bv.StartTime().Add(b.blockSize)) {
 			return
 		}
 
-		res = append(res, bucket.streams(ctx))
+		res = append(res, bv.Streams(ctx, WarmWrite))
+		res = append(res, bv.Streams(ctx, ColdWrite))
 
 		// NB(r): Store the last read time, should not set this when
 		// calling FetchBlocks as a read is differentiated from
 		// a FetchBlocks call. One is initiated by an external
 		// entity and the other is used for streaming blocks between
 		// the storage nodes. This distinction is important as this
 		// data is important for use with understanding access patterns, etc.
-		bucket.setLastRead(b.nowFn())
+		bv.SetLastRead(b.nowFn())
 	})
 
 	return res
 }
 
+func (b *dbBuffer) forEachBucketAsc(fn func(BufferBucketVersions)) {
+	buckets := b.bucketsMap
+	keys := make([]xtime.UnixNano, len(buckets))
+	i := 0
+	for k := range buckets {
+		keys[i] = k
+		i++
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].Before(keys[j])
+	})
```

Comment:
This anon function is going to allocate.

Reply:
Could also just implement the sort interface to avoid the alloc no? Also, just for my own curiosity, how do you know it will alloc? Isn't there a chance the compiler knows it doesn't escape, or does it never do that?

Reply:
So although the func seems to be inlined and does not escape, the call to sort.Slice I believe makes `keys` escape:
```
r@series[juchan/bucket-map]$ go build -gcflags "-m" .  2>&1 | sort | fgrep "buffer.go:4"
./buffer.go:409:14: make([]"github.com/m3db/m3/vendor/github.com/m3db/m3x/time".UnixNano, 0, len(buckets)) escapes to heap
./buffer.go:413:12: keys escapes to heap
./buffer.go:413:19: (*dbBuffer).forEachBucketAsc func literal does not escape
./buffer.go:413:19: can inline (*dbBuffer).forEachBucketAsc.func1
...
```

Reply:
@richardartoul just tried to implement the sort interface, but that solution still makes the thing you pass to `sort.Sort` escape.

I'll keep a separate slice of sorted keys and maintain that.